wgpu/tests/wasm.rs

```diff
@@ -0,0 +1,210 @@
+            entry_point: "vs_main",
+            buffers: &[],
+        },
+        fragment: Some(wgpu::FragmentState {
+            module: &shader,
+            entry_point: "fs_main",
+            targets: &[Some(swapchain_format.into())],
+        }),
+        primitive: wgpu::PrimitiveState::default(),
+        depth_stencil: None,
+        multisample: wgpu::MultisampleState::default(),
+        multiview: None,
+    });
+
+    let config = wgpu::SurfaceConfiguration {
+        usage: wgpu::TextureUsages::RENDER_ATTACHMENT,
+        format: swapchain_format,
+        width: size.width,
+        height: size.height,
+        present_mode: wgpu::PresentMode::Fifo,
+        alpha_mode: surface.get_supported_alpha_modes(&adapter)[0],
+    };
+
+    surface.configure(&device, &config);
+
+    event_loop.spawn(move |event, _, control_flow| {
+        *control_flow = ControlFlow::Wait;
+        if let Event::RedrawRequested(_) = event {
+            let frame = surface
+                .get_current_texture()
+                .expect("Failed to acquire next swap chain texture");
+            let view = frame
+                .texture
+                .create_view(&wgpu::TextureViewDescriptor::default());
+            let mut encoder =
+                device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+            {
+                let mut rpass = encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+                    label: None,
+                    color_attachments: &[Some(wgpu::RenderPassColorAttachment {
+                        view: &view,
+                        resolve_target: None,
+                        ops: wgpu::Operations {
+                            load: wgpu::LoadOp::Clear(wgpu::Color::GREEN),
+                            store: true,
+                        },
+                    })],
+                    depth_stencil_attachment: None,
+                });
+                rpass.set_pipeline(&render_pipeline);
+                rpass.draw(0..3, 0..1);
+            }
+
+            queue.submit(Some(encoder.finish()));
+            frame.present();
+
+            assert_rendering_result(&window);
+
+            *control_flow = ControlFlow::Exit;
+        }
+    });
+}
+
+fn read_pixel(canvas: HtmlCanvasElement, x: i32, y: i32) -> [u8; 4] {
+    let mut result = [0_u8; 4];
+    let context = canvas
+        .get_context("webgl2")
+        .unwrap()
+        .unwrap()
+        .dyn_into::<web_sys::WebGl2RenderingContext>()
+        .unwrap();
+
+    context
+        .read_pixels_with_u8_array_and_dst_offset(
+            x,
+            y,
+            1,
+            1,
+            web_sys::WebGl2RenderingContext::RGBA,
+            web_sys::WebGl2RenderingContext::UNSIGNED_BYTE,
+            &mut result,
+            0,
+        )
+        .unwrap();
+    result
+}
+
+async fn init() -> (Window, EventLoop<()>, Surface, Adapter, Device, Queue) {
+    let event_loop = EventLoop::new();
```

Comment:
I wouldn't bother with winit here - all of our tests run headless, so all we need is the canvas and surface to be created, we don't actually need to interact with the canvas otherwose.